romanisim/util.py

```diff
@@ -204,26 +208,50 @@ def add_more_metadata(metadata):
-    metadata['exposure']['effective_exposure_time'] = effexptime
+
+    if usecrds:
+        # Get the recommendation from the CRDS based on the date and the context
+        rec = crds.getrecommendations({'ROMAN.META.INSTRUMENT.NAME': 'wfi', 'ROMAN.META.EXPOSURE.START_TIME': meta['exposure']['start_time'].value}, observatory='roman')
+        context = api.get_default_context('roman')
+        matab_ref = rec['matable']
+        matab_file = api.dump_references(context, [matab_ref])
+
+        matab = asdf.open(matab_file[matab_ref])
+
+        metadata['exposure']['frame_time'] = matab['roman']['science_tables'][f'SCI{manum:04}']['frame_time']
+        metadata['exposure']['ma_table_id'] = f'sci{manum:04d}'
+        nresultants = len(metadata['exposure']['read_pattern'])
+        metadata['exposure']['nresultants'] = nresultants
+
+        if metadata['exposure']['truncated']:
+            metadata['exposure']['read_pattern'] = matab['roman']['science_tables'][f'SCI{manum:04}']['science_read_pattern'][:nresultants]
+            metadata['exposure']['exposure_time'] = matab['roman']['science_tables'][f'SCI{manum:04}']['accumulated_exposure_time'][:nresultants]  # Subtracting 1 from nresultants due to Python's 0-indexing
+            metadata['exposure']['effective_exposure_time'] = matab['roman']['science_tables'][f'SCI{manum:04}']['effective_exposure_time'][:nresultants]
+        else:
+            metadata['exposure']['read_pattern'] = matab['roman']['science_tables'][f'SCI{manum:04}']['science_read_pattern']
+            metadata['exposure']['exposure_time'] = matab['roman']['science_tables'][f'SCI{manum:04}']['accumulated_exposure_time'][-1]
+            metadata['exposure']['effective_exposure_time'] = matab['roman']['science_tables'][f'SCI{manum:04}']['effective_exposure_time'][-1]
+    else:
+        read_pattern = metadata['exposure'].get(
+            'read_pattern',
+            parameters.read_pattern[metadata['exposure']['ma_table_number']])
+        metadata['exposure']['read_pattern'] = read_pattern
+        openshuttertime = parameters.read_time * read_pattern[-1][-1]
+        offsets = dict(start=0 * u.s, mid=openshuttertime * u.s / 2,
+                    end=openshuttertime * u.s)
+        starttime = metadata['exposure']['start_time']
+        if not isinstance(starttime, Time):
+            starttime = Time(starttime, format='isot')
+        for prefix, offset in offsets.items():
+            metadata['exposure'][f'{prefix}_time'] = Time((
+                starttime + offset).isot)
+        metadata['exposure']['nresultants'] = len(read_pattern)
+        metadata['exposure']['frame_time'] = parameters.read_time
+        metadata['exposure']['exposure_time'] = openshuttertime
+        metadata['exposure']['ma_table_id'] = f'sci{manum:04d}'
+        effexptime = parameters.read_time * (np.mean(read_pattern[-1]))        
+        metadata['exposure']['effective_exposure_time'] = round(effexptime, 4)
```

Comment:
Can we try to pull as much as possible out of this either before or after the if / else?
i.e., metadata.exposure.nresultants I think takes the same path and could be done once instead of duplicated in the if and the else?  Likewise the ma_table_id.  It looks to me like the "else" sets the start / mid / stop time but the "if" does not; presumably we want to set these in both cases?

Ideally I think if I had the logic right in romanisim we could delete the whole "if" case and rely on the "else" case to populate these correctly; maybe you and I should chat about that option at some point.